darshan-runtime/lib/darshan-ldms.c

```diff
@@ -0,0 +1,270 @@
+        }
+}
+
+ldms_t setup_connection(const char *xprt, const char *host,
+                        const char *port, const char *auth)
+{
+        const char *timeout = "5";
+        int rc;
+        struct timespec ts;
+
+        if (!timeout) {
+                ts.tv_sec = time(NULL) + 5;
+                ts.tv_nsec = 0;
+        } else {
+                int to = atoi(timeout);
+                if (to <= 0)
+                        to = 5;
+                ts.tv_sec = time(NULL) + to;
+                ts.tv_nsec = 0;
+        }
+
+        ldms_g = ldms_xprt_new_with_auth(xprt, auth, NULL);
+        if (!ldms_g) {
+                printf("Error %d creating the '%s' transport\n",
+                       errno, xprt);
+                return NULL;
+        }
+
+        sem_init(&dC.recv_sem, 1, 0);
+        sem_init(&dC.conn_sem, 1, 0);
+
+        rc = ldms_xprt_connect_by_name(ldms_g, host, port, event_cb, NULL);
+        if (rc) {
+                printf("Error %d connecting to %s:%s\n",
+                       rc, host, port);
+                return NULL;
+        }
+        sem_timedwait(&dC.conn_sem, &ts);
+        if (dC.conn_status)
+                return NULL;
+        return ldms_g;
+}
+
+void darshan_ldms_connector_initialize(struct darshan_core_runtime *init_core)
+{
+     /*TODO: Create environment variable to re-connect to ldms every x seconds
+        if(getenv("DARSHAN_LDMS_REINIT"))
+            dC.env_ldms_reinit = getenv("DARSHAN_LDMS_REINIT");
+        else
+            dC.env_ldms_reinit = "1";*/
+     
+    (void)gethostname(dC.hname, sizeof(dC.hname));
+    
+    dC.uid = init_core->log_job_p->uid;
+    
+    if (getenv("SLURM_JOB_ID"))
+        dC.jobid = atoi(getenv("SLURM_JOB_ID"));
+    else if (getenv("LSB_JOBID"))
+    	dC.jobid = atoi(getenv("LSB_JOBID"));
+    else if (getenv("JOB_ID"))
+        dC.jobid = atoi(getenv("JOB_ID"));
+    else if (getenv("LOAD_STEP_ID"))
+        dC.jobid = atoi(getenv("LOAD_STEP_ID"));
+    else
+    /* grab jobid from darshan_core_runtime if slurm, lsf, sge or loadleveler do not exist*/
+        dC.jobid = init_core->log_job_p->jobid;
+    
+    /* grab exe path from darshan_core_runtime */
+    dC.exename = strtok(init_core->log_exemnt_p, " ");
+     
+    /* Pull executable name from proc if no arguemments are given. */
+    if (dC.exename == NULL)
+    {
+         char buff[DARSHAN_EXE_LEN];
+         int len = readlink("/proc/self/exe", buff, sizeof(buff)-1);
+         buff[len] = '\0';
+         dC.exename = buff;
+    }
+
+    /* Set flags for various LDMS environment variables */
+    if (getenv("POSIX_ENABLE_LDMS"))
+        dC.posix_enable_ldms = 0;
+    else
+        dC.posix_enable_ldms = 1;
+
+    if (getenv("MPIIO_ENABLE_LDMS"))
+        dC.mpiio_enable_ldms = 0;
+    else
+        dC.mpiio_enable_ldms = 1;
+
+    /* Disable STDIO if verbose is enabled to avoid a recursive
+    function for darshan_ldms_connector_send() */
```

Comment:
You're right, I was seeing a deadlock. Any ``printf`` statement in `darshan_ldms_connector_send()`` or any of the macros would trigger Darshan of a "write" which would call these functions again and thus, output the same ``printf`` statement which would trigger Darshan and the cycle would continue.  The only way to stop this was for the job to time out or I would have to cancel it. 

I have replaced the ``printf`` with ``darshan_core_fprintf()`` and it works! Thanks for the information!

Reply:
Ahh, that makes sense. We've definitely seen that kind of infinite loop behavior behavior before when we're accidentally calling the wrapped functions instead of the actual implementation. As you point out, it just basically turns into an endless loop of invoking the wrapper function instead of eventually calling the real implementation.

Glad it was a simple fix!